components/backup-stream/src/tempfiles.rs

```diff
@@ -0,0 +1,1003 @@
+            return Err(Error::new(
+                ErrorKind::Other,
+                "open_for_write isn't allowed when there are concurrent reading.",
+            ));
+        }
+        let fr = ForWriteCore {
+            core: Arc::clone(&f.content),
+            ref_counter: Arc::clone(&f.writer_count),
+            rel_path: p.to_owned(),
+            done_result: None,
+        };
+        f.writer_count.fetch_add(1, Ordering::SeqCst);
+        match self.cfg.content_compression {
+            CompressionType::Unknown => Ok(ForWrite::Plain(fr)),
+            CompressionType::Zstd => Ok(ForWrite::ZstdCompressed(ZstdCompressionWriter::new(fr))),
+            unknown_compression => Err(Error::new(
+                ErrorKind::Unsupported,
+                format!(
+                    "the compression method {:?} isn't supported for now.",
+                    unknown_compression
+                ),
+            )),
+        }
+    }
+
+    /// Open a file reference for reading.
+    /// Please notice that once a compression applied, this would yield the
+    /// compressed content (won't decompress them.) -- that is what "raw"
+    /// implies.
+    /// "But why there isn't a `open_for_read` which decompresses the content?"
+    /// "Because in our use case, we only need the raw content -- we just send
+    /// it to external storage."
+    pub fn open_raw_for_read(&self, p: &Path) -> std::io::Result<ForRead> {
+        use std::io::{Error, ErrorKind};
+
+        let fs = self.files.lock().unwrap();
+        let f = fs.items.get(p);
+        if f.is_none() {
+            return Err(Error::new(
+                ErrorKind::NotFound,
+                format!("file {} not found", p.display()),
+            ));
+        }
+        let f = f.unwrap();
+        let refc = f.writer_count.load(Ordering::SeqCst);
+        if refc > 0 {
+            // NOTE: the current implementation doesn't allow us to write when there are
+            // readers, because once the writter swapped out the file, the reader may not
+            // notice that. Perhaps in the future, we can implement something
+            // like cursors to allow the reader be able to access consistent
+            // File snapshot even there are writers appending contents
+            // to the file. But that isn't needed for now.
+            return Err(Error::new(
+                ErrorKind::Other,
+                format!(
+                    "open_for_read isn't allowed when there are concurrent writing (there are still {} reads for file {}.).",
+                    refc,
+                    p.display()
+                ),
+            ));
+        }
+        let st = f.content.lock().unwrap();
+        let myfile = if st.external_file.is_some() {
+            Some(self.open_relative(p)?)
+        } else {
+            None
+        };
+        Ok(ForRead {
+            content: Arc::clone(&f.content),
+            myfile,
+            ref_counter: Arc::clone(&f.reader_count),
```

Comment:
when to increase `f.reader_count`?

BTW can we use the same name of `ref_counter` and `xxx_ count`?